docs/source/examples/batch-inference.rst

```diff
@@ -0,0 +1,315 @@
+
+The following is an example script, where we aggregate multiple inputs into a single batch for better GPU utilization, and process the entire chunk of data batch by batch:
+
+.. code-block:: python
+    
+    from vllm import LLM
+    
+    BATCH_CHAR_COUNT = 2000
+    DATA_PATH = '/data/part_0.jsonl'
+    OUTPUT_PATH = '/output'
+
+    llm = LLM(model='meta-llama/Meta-Llama-3.1-7B-Instruct', tensor_parallel_size=1)
+
+    def batch_inference(llm: LLM, data_path: str):
+        # This can take about 1-2 hours on a L4 GPU.
+        print(f'Processing {data_path}...')
+        data_name = data_path.split('/')[-1]
+
+        # Read data (jsonl), each line is a json object
+        with open(data_path, 'r') as f:
+            data = f.readlines()
+            # Extract the first message from the conversation
+            messages = [json.loads(d.strip())['conversation'][0]['content'] for d in data]
+
+        # Run inference
+        batch_char_count = 0
+        batch_messages = []
+        generated_text = []
+        for message in tqdm(messages):
+            # Calculate the word count of the conversation
+            char_count = len(message)
+            batch_char_count += char_count
+
+            if batch_char_count > BATCH_CHAR_COUNT:
+                outputs = llm.generate(batch_messages, SAMPLING_PARAMS, use_tqdm=False)
+                generated_text = []
+                for output in outputs:
+                    generated_text.append(' '.join([o.text for o in output.outputs]))
+                batch_messages = []
+                batch_char_count = 0
+
+            batch_messages.append(message)
+
+        # Save predictions
+        os.makedirs(OUTPUT_PATH, exist_ok=True)
+        with open(os.path.join(OUTPUT_PATH, data_name), 'w') as f:
+            for text in generated_text:
+                f.write(text + '\n')
+
+    batch_inference(llm, DATA_PATH)
+
+For complete script, see `examples/batch_inference/inference.py <https://github.com/skypilot-org/skypilot/blob/main/examples/batch_inference/inference.py>`_ and you can run it with ``HF_TOKEN=<your-huggingface-token> python inference.py`` to test it on the dev machine.
+
+After testing it on the dev machine, we can now compose a task yaml (`inference.yaml <https://github.com/skypilot-org/skypilot/blob/main/examples/batch_inference/inference.yaml>`) to run the inference on clouds.
+
+.. code-block:: bash
+
+    # Set HuggingFace token for accessing Llama model weights.
+    export HF_TOKEN=...
+    sky launch -c inf ./inference.yaml \
+        --env HF_TOKEN
+
+.. TODO: make r2 bucket publically accessible
+.. tested with inference.py and inference.yaml on 2024-09-15 and works well.
+
+.. _scale-out-to-multiple-nodes:
+
+Scale Out to Multiple Nodes
+---------------------------
+
+To scale out the inference to multiple machines, we can group the data chunks into multiple pieces so that each machine can process one piece.
```

Comment:
Reduce concept: yet another one here 'piece'!

Reply:
Removed this concept now.